col/src/main/java/vct/col/ast/util/NameScanner.scala

```diff
@@ -60,15 +86,39 @@ class NameScanner extends RecursiveVisitor[AnyRef](null, null) {
     }
   }
 
+  /** If the name is in `additionalNames`, or there is a regular declaration with the same name, the types must match
+    * If there is a free variable with the same name, the types must match
+    * If there is no decl nor a free variable, we found a new free variable, which is added to the `freeNames` list
+    *
+    * @param additionalNames can be used to pass in given or yields scopes.
+    * @return true if `name` is the name of a free variable.
+    */
+  private def checkName(name: String, typ: Type, additionalNames: Function1[String, Option[DeclarationStatement]]): Boolean =
+    (additionalNames(name) orElse getDecl(name), freeNames.get(name)) match {
+      case (Some(decl), _) =>
+        if (typ != null && decl.`type` != typ) {
```

Comment:
You may prefer `Option(typ)`, which is equivalent to `match typ { null => None; other => Some(other) }`

Reply:
and then `Option(typ).isEmpty` or whatever, but it may just be more verbose, idk.

Reply:
I am not sure how Option(typ) improves it, I guess it can replace `typ != null`? But I think that doesn't make it much clearer, so if its okay with you I will keep it as is.